Fix #138: Topic train fragment Low-fi UI (Part 4)

rt4914 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Replicated from #197 

Part 1: #200 
Part 2: #202 
Part 3: #203
Part 4: Contains test case for TopicTrainFragment

Mock: https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/f403add1-079a-4202-9975-efe04c076290/Home-Page-1-Start-Learning-Pre-Scroll-6

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

Commit:
All xml files for topic-train except layout folder

Commit:
Skill item related files

Commit:
TopicTrainFragment implementation

Commit:
Topic Train Test Case

rt4914 commented:
@veena14cs @nikitamarysolomanpvt PTAL

veena14cs reviewed (approved):
LGTM

rt4914 commented:
> LGTM

Thanks @veena14cs

nikitamarysolomanpvt commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -26,9 +31,13 @@ import javax.inject.Singleton
 class TopicTrainFragmentTest {
 
   @Test
```
Please include KDoc for test cases

BenHenning replied:
Is there specific documentation you're looking for @nikitamarysolomanpvt? Typically we don't document tests (mostly because they should be sufficiently self-explanatory), but sometimes it's useful.

nikitamarysolomanpvt replied:
Initially I was using autogenerated test case method names in android studio.I had gone through your slide on test case then understood the new approach, the approach is quite nice and as per that we could understand the test and expected result by the method name. I think we could just give the link of the standards (ppt which explain this) that we are following inside doc (for new contributors  to understand)and avoid self-explanatory method KDoc.

nikitamarysolomanpvt reviewed (commented):
Please include KDoc for test cases

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -26,9 +31,13 @@ import javax.inject.Singleton
-  fun testTopicTrainFragment_loadFragment_textIsDisplayed() {
-    ActivityScenario.launch(TopicActivity::class.java).use {
-      onView(withId(R.id.dummy_text_view)).check(matches(withText("This is dummy TextView for testing")))
+  fun testTopicTrainFragment_loadFragment_selectSkills_submitButtonIsActivated() {
```
Other cases worth testing:
- Verifying that when no skills are selected, the button is not clickable
- Verifying that the button's clickable state becomes disabled after selecting & deselecting skills
- Verifying that clicking the submit button properly forwards the list of skills to an outgoing activity
- Verifying that the list of selected skills stay selected upon a configuration change
- Verifying that the submit button activation status stays correct for both enabled/disabled cases upon a configuration change

rt4914 replied:
Added all these working test-cases

Commit:
Introduce QuestionPlayer Activity/Fragment

Commit:
Domain layer code integration

Commit:
Resolve merge conflict develop

Commit:
Merge remote-tracking branch 'upstream/topic-train-low-fi-part-3' into topic-train-low-fi-part-4

Commit:
Start button activated

Commit:
Start button activated

Commit:
Separate ViewModel for item

Commit:
Managing configuration changes

Commit:
Checkbox state save

Commit:
Nit changes

Commit:
Fixing HomeActivityController

Commit:
Managing configuration change in checkboxes

Commit:
Proper route to QuestionPlayer

Commit:
Merge remote-tracking branch 'upstream/topic-train-low-fi-part-3' into topic-train-low-fi-part-4

Commit:
QuestionPlayerActivity test case

Commit:
Nit changes

Commit:
Merge remote-tracking branch 'upstream/topic-train-low-fi-part-3' into topic-train-low-fi-part-4

Commit:
More test cases

Commit:
Nit changes

Commit:
Resolve merge conflicts

Commit:
Merge remote-tracking branch 'upstream/topic-train-low-fi-part-3' into topic-train-low-fi-part-4

rt4914 commented:
@BenHenning PTAL

Commit:
EOF added

BenHenning commented on app/src/sharedTest/java/org/oppia/app/RecyclerViewMatcher.kt:
```diff
@@ -0,0 +1,54 @@
+import org.hamcrest.Matcher
+import org.hamcrest.TypeSafeMatcher
+
+class RecyclerViewMatcher(private val recyclerViewId: Int) {
```
Could we maybe use the one in BindableAdapterTest, instead? Prefer extracting that one to here for reuse since I think that one has all the necessary functionality, and is simpler.

Also, please add documentation and credit links where applicable (if you didn't derive this solution on your own).

rt4914 replied:
I was not able to use code from `BindableAdapterTest `, but I have updated the code for better readability.

rt4914 replied:
Also, I have created a custom adapter for this TopicTrainFragment, so will the `BindableAdapterTest ` work with this?

BenHenning commented on app/src/sharedTest/java/org/oppia/app/RecyclerViewMatcher.kt:
```diff
@@ -0,0 +1,54 @@
+    return atPositionOnView(position, -1)
+  }
+
+  fun atPositionOnView(position: Int, targetViewId: Int): Matcher<View> {
```
This method should have brief documentation to explain how to use it.

rt4914 replied:
Added brief explanation of this method.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/RecyclerViewMatcher.kt:
```diff
@@ -0,0 +1,54 @@
+import org.hamcrest.Matcher
+import org.hamcrest.TypeSafeMatcher
+
+class RecyclerViewMatcher(private val recyclerViewId: Int) {
```
Also, do we need the wrapper RecyclerViewMatcher? Could we just have a top-level Kotlin file with the atPosition method and import it where needed? The BindableAdapterTest version lends itself to this approach since it follows the standard Espresso matcher idiom of being a static-esque, standalone method.

rt4914 replied:
Removed wrapper and started using `atPosition` and `atPositionView` method directly.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -2,36 +2,138 @@ package org.oppia.app.topic.train
-      onView(withId(R.id.dummy_text_view)).check(matches(withText("This is dummy TextView for testing")))
+      onView(withId(R.id.master_skills_text_view)).check(matches(withText(R.string.topic_train_master_these_skills)))
+      onView(withRecyclerView(R.id.skill_recycler_view).atPosition(0)).perform(click())
+      onView(withId(R.id.topic_train_start_button)).check(matches(isClickable()))
```
This seems to be testing several things--I suggest splitting it up. If you ensure each test follows distinct arrange/act/assert blocks (versus arrange/act/assert/act/assert) then the tests will be a bit cleaner.

You can also assume previous tests pass, and use that as an assumption for certain state being correct at specific points in the test. E.g. for this case, you can have a test that verifies selecting skills works (that they are selected), another that verifies that selecting skills properly enables the start button, a third that verifies selecting then deselecting skills (with no assertion in between) properly deselects, and a fourth that ensures selection + deselection properly disables the start button.

You may notice that when splitting this into four tests, there are others that become apparent and maybe worthwhile to test: selection/deselecting different numbers of skills to verify that all skills must be selected for the button to become inactive, but only one skill needs to be selected in order for it to be active.

I suggest adding some more tests around these areas to ensure the checkboxes & button states are well tests.

rt4914 replied:
Updated test-cases to be smaller and precise.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -2,36 +2,138 @@ package org.oppia.app.topic.train
+    onView(withRecyclerView(R.id.skill_recycler_view).atPosition(1)).perform(click())
+    onView(withId(R.id.topic_train_start_button)).check(matches(isClickable()))
+    onView(withId(R.id.topic_train_start_button)).perform(click())
+    intended(hasComponent("org.oppia.app.topic.questionplayer.QuestionPlayerActivity"))
```
Could we depend directly on QuestionPlayerActivity's class & to retrieve the intent extra KEY? Otherwise refactoring may be tricky, or renaming the intent key could break this test.

rt4914 replied:
Resolved. Used QuestionPlayerActvity name and key directly from the QuestionPlayerActvity file.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -2,36 +2,138 @@ package org.oppia.app.topic.train
+  @Before
+  fun setUp() {
+    Intents.init()
+    skillIdListIntent.add("test_skill_id_0")
```
Maybe just ``skillIdList`` would suffice since this isn't technically an intent, but it is a payload to one.

rt4914 replied:
Updated to `skillIdList`

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -2,36 +2,138 @@ package org.oppia.app.topic.train
+        0,
+        R.id.skill_check_box
+      )
+    ).check(matches(isChecked()))
```
Do we need to recheck it since this is verified above in its own test? We can assume that test is passing when we get here.

rt4914 replied:
It was just extra added layer of testing but definitely if we remove this than also it will work.

Final: Removed checking `checkbox` state.

BenHenning reviewed (approved):
Thanks @rt4914!

Commit:
Resolve merge conflicts

Commit:
Test nit changes

Commit:
Merge remote-tracking branch 'upstream/develop' into topic-train-low-fi-part-4

Commit:
Test case updated

nikitamarysolomanpvt reviewed (approved):
LGTM

Commit:
Merge remote-tracking branch 'upstream/develop' into topic-train-low-fi-part-4

Commit:
Updated RecyclerViewMatcher

Commit:
Nit change

rt4914 commented:
@BenHenning PTAL at updated test cases.

BenHenning commented on app/src/main/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivity.kt:
```diff
@@ -27,5 +27,9 @@ class QuestionPlayerActivity : InjectableAppCompatActivity() {
       return intent
     }
+
+    fun getIntentKey():String{
```
Nit: please put spacing between elements (e.g. ``fun getIntentKeys(): String {``

rt4914 replied:
Done.

Commit:
Test added

rt4914 commented:
@BenHenning Regarding RecyclerViewMatcher -> I have removed that file and implemented those methods inside BindableTestAdapter.

I was getting stuck at two points.

1. I wanted to perform click on `Checkbox` inside the item which I was not unable do with the existing `BindableTestAdapter`.
2. I wanted the information of `Checkbox` inside the item too.

So, I have added two methods inside `BindableTestAdapter `which solves the issue that i am facing.

Commit:
Merge remote-tracking branch 'upstream/develop' into topic-train-low-fi-part-4

BenHenning reviewed (changes requested):
Thanks @rt4914. I ran out of time to deep dive this PR, but could you move the test utilities you're using in BindableAdapterTest to a shared location for all tests to use? Prefer not having one test depend on another.

BenHenning commented:
Please reassign to me once the requested refactor is completed.

rt4914 commented:
> Thanks @rt4914. I ran out of time to deep dive this PR, but could you move the test utilities you're using in BindableAdapterTest to a shared location for all tests to use? Prefer not having one test depend on another.

@BenHenning  I am actually confused what you mean here?

My interpretation: I should move the `atPosition` and `atPositionView` functions that I have introduced in BindableTestAdapter to a separate file.

If this is the case, I had actually done that only earlier by introducing the RecyclerViewMatcher file.
Reference Code: https://github.com/oppia/oppia-android/blob/f863f6e386882d4daed8e8cd142dd994efd7eec1/app/src/sharedTest/java/org/oppia/app/recyclerview/RecyclerViewMatcher.kt#L11

Commit:
Merge remote-tracking branch 'upstream/develop' into topic-train-low-fi-part-4

Commit:
Nit changes

Commit:
Nit changes

Commit:
Merge remote-tracking branch 'upstream/develop' into topic-train-low-fi-part-4

Commit:
Test cases

rt4914 commented:
@BenHenning 
As discussed in one of our meetings, for recyclerview item testing we can either introduce `androidx.test.espresso:espresso-contrib:3.1.0`  library or we can write our own custom recyclerview test code.

I tried using `androidx.test.espresso:espresso-contrib:3.1.0` but it only provides code to click on an specific item in recyclerview. There is no mechanism to check the content of an item.

Because of these reasons, I have created one new file which we can use for recyclerview testing, i.e., `RecyclerViewMatcher`, also, I have changed your `BindableAdapterTest` cases to use this `RecyclerViewMatcher` file.

rt4914 commented:
@BenHenning This PR is ready for your review.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -2,36 +2,139 @@ package org.oppia.app.topic.train
+  fun testTopicTrainFragment_loadFragment_displaySkills_startButtonIsInactive() {
+    ActivityScenario.launch(TopicActivity::class.java).use {
+      onView(withId(R.id.master_skills_text_view)).check(matches(withText(R.string.topic_train_master_these_skills)))
+      onView(atPosition(R.id.skill_recycler_view,0)).check(matches(hasDescendant(withId(R.id.skill_check_box))))
```
Nit: please put a space between ``,`` and ``0``. Ditto elsewhere in this file.

rt4914 replied:
Resolved in this file and BindabaleTestAdapter

BenHenning commented on app/src/sharedTest/java/org/oppia/app/topic/train/TopicTrainFragmentTest.kt:
```diff
@@ -2,36 +2,139 @@ package org.oppia.app.topic.train
+  fun testTopicTrainFragment_loadFragment_selectSkills_configurationChange_skillsAreSelected() {
+    activityTestRule.launchActivity(null)
+    onView(atPosition(R.id.skill_recycler_view,0)).perform(click())
+    activityTestRule.activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
```
I don't believe this is sufficient for a configuration change. Suggest instead: ``ActivityScenario.recreate()``

Ditto elsewhere for configuration changes.

rt4914 replied:
This means I have add this or do I have to remove the current line and replace this one?

rt4914 replied:
Using both and merge code for now, because the test-cases passes in all possible combinations, using current code only, using your code only and using both.

BenHenning reviewed (approved):
Thanks @rt4914! If the tests are passing, this looks reasonable to me. I found one potential issue with some of the new tests around configuration changes, but other than that the PR LGTM. Feel free to submit once the comments are resolved.

Commit:
Test case updated

rt4914 commented:
Merging this code after making all suggested changes.
